code/parachain/frame/staking-rewards/src/lib.rs

```diff
@@ -1710,76 +1694,35 @@ pub(crate) fn do_reward_accumulation<T: Config>(
-			reward.max_rewards.into()
-		{
-			let balance_on_hold = T::Assets::balance_on_hold(asset_id, pool_account);
-			if !balance_on_hold.is_zero() && balance_on_hold >= newly_accumulated_rewards.into() {
-				T::Assets::release(
-					asset_id,
-					pool_account,
-					newly_accumulated_rewards.into(),
-					false, // not best effort, entire amount must be released
-				)
-				.expect("funds should be available to release based on previous check; qed;");
-
-				reward.total_rewards = new_total_rewards.into();
-				reward.last_updated_timestamp = last_updated_timestamp;
-				Ok(())
-			} else {
-				Err(RewardAccumulationCalculationError::RewardsPotEmpty)
-			}
-		} else if reward.total_rewards.saturating_add(reward.total_dilution_adjustment) <
-			reward.max_rewards
-		{
-			// if the new total rewards are less than or equal to the max rewards AND the current
-			// total rewards are less than the max rewards (i.e. the newly accumulated rewards is
-			// less than the the amount that would be accumulated based on the periods surpassed),
-			// then release *up to* the max rewards
-
-			// REVIEW(benluelo): Should max_rewards be max_periods instead? Currently, the
-			// max_rewards isn't updatable, and once the max_rewards is hit, it's expected that no
-			// more rewards will be accumulated, so it's ok to not reward an entire period's worth
-			// of rewards. Review this if the max_rewards ever becomes updateable in the future.
-
-			// SAFETY(benluelo): Usage of Sub::sub: reward.total_rewards is known to be less than
-			// reward.max_rewards as per check above
-			let rewards_to_release = reward.max_rewards.sub(reward.total_rewards).into();
-
-			let balance_on_hold = T::Assets::balance_on_hold(asset_id, pool_account);
-			if !balance_on_hold.is_zero() && balance_on_hold >= newly_accumulated_rewards.into() {
-				T::Assets::release(
-					asset_id,
-					pool_account,
-					rewards_to_release.into(),
-					false, // not best effort, entire amount must be released
-				)
-				.expect("funds should be available to release based on previous check; qed;");
-
-				// return an error, but update the reward first
-				reward.total_rewards = reward.max_rewards;
-				reward.last_updated_timestamp = last_updated_timestamp;
-				Err(RewardAccumulationCalculationError::MaxRewardsAccumulated)
-			} else {
-				Err(RewardAccumulationCalculationError::RewardsPotEmpty)
+		if !total_locked_rewards.is_zero() {
+			T::Assets::release(
+				asset_id,
+				pool_account,
+				newly_accumulated_rewards.into(),
+				false, // not best effort, entire amount must be released
+			)
+			// Note: will this always be Ok or need to handle error?
+			.expect("funds should be available to release based on previous check; qed;");
+
+			reward.total_rewards = new_total_rewards.into();
+			reward.last_updated_timestamp = last_updated_timestamp;
+
+			let new_balance_on_hold = T::Assets::balance_on_hold(asset_id, pool_account);
+
+			if new_balance_on_hold.into().is_zero() {
```

Comment:
you should check that it's below the reward rate for one period, otherwise it could end up in a situation where the amount is less than required to accumulate one period but it still attempts to accumulate every block (which would be wasteful)

Reply:
updating the reward pot empty storage is now being done in two places - here and in `reward_accumulation_hook_reward_update_calculation`. It should be removed from one of them.

since `do_reward_accumulation` doesn't currently read the `RewardPotIsEmpty` storage, I don't think it should write to it either. It just accumulates if it can, and reports the relevant errors which are then handled at the callsite.